keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Backwards compatibility
+
+There is a backwards compatibility issue involved when tightening the validation on snapshot objects. Since the feature is already in beta we are committed to more backward compatibility guarantee than alpha.
+
+There are two perspectives.
+- client perspective: API requests that previously succeeded now fail (these are okay)
+  - create: invalid things get rejected rather than persisted
+  - update: invalid transitions now get rejected
+- API server perspective for existing persisted data
+  - do not tighten validation (via schema or webhook) in a way that would prevent a no-op update (this must not be violated)
+  - reason: removing a finalizer is done via update, causes problems with delete
+  - making a previously optional field required in the schema blocks update of previously persisted data that omitted the field (unless the update populates the newly required field, or you specify a schema default)
+
+To tackle the backwards compatibility problem, this KEP proposes the following release process.
+
+Begin with validating webhook only enforcement. The webhook will perform the following validation
+- one release with ratcheting validation using the webhook server
+   - webhook is strict on create
+   - webhook is strict on updates where the existing object passes strict validation
+   - webhook is relaxed on updates where the existing object fails strict validation (allows finalizer removal, status update, deletion, etc)
+ - some reconciliation process in volume snapshot controller (this is the hard part) that ensures invalid data is fixed or removed
+
+Once we are sure no invalid data is persisted, we can switch to CRD schema-enforced validation with validating webhooks for immutability in a subsequent release.
+
+We are unsure of the specifics of the reconciliation process in which the controller will remove or fix invalid objects.
+
+#### Current Controller validation of OneOf semantic
+
+##### Handling [VolumeSnapshot](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller.go#L192).
+
+If the object violates oneOf semantic: Update the VS status to “SnapshotValidationError” and issue an event.
+
+Note:
+- If the VS object has been updated AFTER binding to a VSC, binding from VS->VSC will be lost.
+- Deletion of an invalid resource is not blocked by that check as the deletion workflow happens before validation(code). This is to ensure that a user can delete an invalid VS resource.
+
+##### Handling [VolumeSnapshotContent](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller.go#L91)
```

Comment:
Same here.  Please reference from a release branch.

Reply:
I've used the release tag 2.1.1